gap/singular.gi

```diff
@@ -0,0 +1,123 @@
+#############################################################################
+##
+## variables and polynomials are a problem - GAP is very slow in computing
+## with polynomials and Singular is fast, but has bugs.
+##
+
+IsSingularAvailable_known:= false;
+
+if not IsBound( StartSingular ) then
+  # Avoid warnings about unbound global variables.
+  StartSingular:= "dummy";
+fi;
+
+BindGlobal( "StartSingularIfAvailable", function()
+  local IsExec, sing_exec;
+
+  # Once we got a positive answer, we need not check.
+  if IsSingularAvailable_known = true then
+    return true;
+  fi;
+
+  # If the Singular package is not available
+  # then we cannot use the Singular executable.
+  if not IsPackageMarkedForLoading( "Singular", "" ) then
+    return false;
+  fi;
+
+  # Use the code from Singular's
+  # 'CheckSingularExecutableAndTempDir' in order to determine
+  # whether a Singular executable will be found by the Singular package.
+  # Note that the fact that the Singular package had been loaded
+  # does not imply that a Singular executable will be found.
+  IsExec:= path -> IsString( path ) and not IsDirectoryPath( path )
+                   and IsExecutableFile( path );
+
+  if IsBoundGlobal( "sing_exec" ) then
+    # The Singular package has been loaded.
+    sing_exec:= ValueGlobal( "sing_exec" );
+    if not IsString( sing_exec ) then
+      # This may happen after a failed 'StartSingular()' call.
+      sing_exec:= "singular";
+    fi;
+  else
+    # Use the default path.
+    sing_exec:= "singular";
+  fi;
+  if IsDirectoryPath( sing_exec ) then
+    sing_exec:= Filename( Directory( sing_exec ), "Singular" );
+  elif not IsExecutableFile( sing_exec ) then
+    sing_exec:= Filename( DirectoriesSystemPrograms(), sing_exec );
+  fi;
+  if not IsExec( sing_exec ) then
+    sing_exec:= Filename( DirectoriesSystemPrograms(), "Singular" );
+  fi;
+  if IsExec( sing_exec ) then
+    IsSingularAvailable_known:= true;
+    StartSingular();
+    return true;
+  fi;
+  return false;
+end );
+
+if IsString( StartSingular ) then
+  Unbind( StartSingular );
+fi;
+
+StartSingularIfAvailable();
+ORDER := MonomialGrlexOrdering();
+
+if not IsBound( SINGULARGBASIS ) then
+  SINGULARGBASIS:= "dummy";
+fi;
+
+BindGlobal( "CallGroebner", function( list )
+    local oldvar, new;
+    if Length(list) = 0 then return list; fi;
+    if StartSingularIfAvailable() then
```

Comment:
Hmm, is `StartSingular` reentrant? I.e. if we already called it, does it start another Singular, or does it detect one is already running and silently returns?

Reply:
The idea is that `StartSingularIfAvailable` sets the variable `IsSingularAvailable_known` to `true` as soon as the standalone Singular can be started. (Well, I think currently we cannot really detect that starting Singular has actually worked.) And as soon as `IsSingularAvailable_known` is `true`, further calls to `StartSingularIfAvailable` will not start a new Singular.

Thinking now again about this, we have a problem when Singular's `CloseSingular` gets called, because then `IsSingularAvailable_known` will still be `true`.

(Is there a solution for LiePRing without expecting changes in the Singular package? 
What about checking `IsStream( Sing_Proc ) and not IsClosedStream( Sing_Proc )` in `StartSingularIfAvailable`?)